farm/data_handler/processor.py

```diff
@@ -1671,6 +1672,168 @@ def _sample_to_features(self, sample) -> dict:
+                            "hard_negative_ctxs": List[{'title': str, text': str, 'score': int, 'title_score': int, 'passage_id': int}]
+                            }
+        Returns:
+                sample: instance of Sample
+        """
+        # extract query, positive context passages and titles, hard-negative passages and titles
+        query = self._normalize_question(dictionary["question"])
+        positive_ctx_text = dictionary["positive_ctxs"][self.num_positives]["text"]
+        positive_ctx_title = dictionary["positive_ctxs"][self.num_positives].get("title", None)
+        hard_negative_ctx_texts = [passage["text"] for passage in dictionary["hard_negative_ctxs"][:self.num_hard_negatives]]
+        hard_negative_ctx_titles = [passage.get("title", None) for passage in dictionary["hard_negative_ctxs"][:self.num_hard_negatives]]
+
+        # all context passages and labels: 1 for positive context and 0 for hard-negative context
+        all_ctxs = [positive_ctx_text] + hard_negative_ctx_texts
+        ctx_label = [1]*self.num_positives + [0]*self.num_hard_negatives
+
+        # tokenize query and contexts
+        tokenized_query = tokenize_with_metadata(query, self.query_tokenizer)
+        tokenized_positive_ctx_text = tokenize_with_metadata(positive_ctx_text, self.passage_tokenizer)
+        tokenized_positive_ctx_title = tokenize_with_metadata(positive_ctx_title, self.passage_tokenizer)
+        tokenized_hard_negative_ctx_text = [tokenize_with_metadata(ctx, self.passage_tokenizer) for ctx in hard_negative_ctx_texts]
+        tokenized_hard_negative_ctx_titles = [tokenize_with_metadata(ctx, self.passage_tokenizer) for ctx in hard_negative_ctx_titles]
+
+        if len(tokenized_query["tokens"]) == 0 or tokenized_positive_ctx_text["tokens"] == 0 or len(tokenized_hard_negative_ctx_text["tokens"]) == 0:
+            logger.warning(f"The text could not be tokenized, likely because it contains a character that the tokenizer does not recognize")
+            return None
+
+        clear_text = {"query_text": query,
+                      "positive_ctx_text": positive_ctx_text,
+                      "positive_ctx_title": positive_ctx_title,
+                      "hard_negative_ctx_text": hard_negative_ctx_texts,
+                      "hard_negative_ctx_title": hard_negative_ctx_titles,
+                      "all_ctxs": all_ctxs,
+                      "label": ctx_label
+                     }
+
+        tokenized = {"positive_ctx_tokens": tokenized_positive_ctx_text,
+                     "positive_ctx_title_tokens": tokenized_positive_ctx_title,
+                     "hard_negative_ctx_tokens": tokenized_hard_negative_ctx_text,
+                     "hard_negative_ctx_titles_tokens": tokenized_hard_negative_ctx_titles,
+                     "query_tokens": tokenized_query,
+                     "label_tokens": ctx_label
```

Comment:
don't think we need "label_tokens" here as it won't get tokenized

Reply:
removed it